docs: New docs for table functions (#3727)
confluentinc/ksql: docs/developer-guide/udf.rst

    public static Udaf<String, Struct, Double> averageStringLength(final String initialString){...}

.. _ksql-udtfs:

UDTFs
-----

To create a UDTF you need to create a class that is annotated with ``@UdtfDescription``.
Each method in the class that represents a UDTF must be public and annotated with ``@Udtf``. The class
you create represents a collection of UDTFs all with the same name but may have different
arguments and return types.

``@UdfParameter`` annotations can be added to method parameters to provide users with richer
information, including the parameter schema. This annotation is required if the KSQL type cannot
be inferred from the Java type (e.g. ``STRUCT``).

Null Handling
~~~~~~~~~~~~~

> If a UDTF uses primitive types in its signature it is indicating that the parameter should never be null.

Reviewer: ```suggestion
If a UDTF uses primitive types in its signature it is indicating that the parameter should never be null.
```
```suggestion
If a UDTF uses primitive types in its signature, this indicates that the parameter should never be null.
```

Author: this is just copy and paste from the UDF section :)

> Conversely, using boxed types indicates the function can accept null values for the parameter.

Reviewer: ```suggestion
Conversely, using boxed types indicates the function can accept ``null`` values for the parameter.
```

It is up to the implementor of the UDTF to chose which is the most appropriate.
A common pattern is to return ``null`` if the input is ``null``, though generally this is only for
parameters that are expected to be supplied from the source row being processed. For example,
a ``substring(String str, int pos)`` UDF might return null if ``str`` is null, but a
null ``pos`` parameter would be treated as an error, and hence should be a primitive.

> (In actual fact, the in-built substring is more lenient and would return null if pos was null).

Reviewer: ```suggestion
The in-built ``substring`` is more lenient and returns null if ``pos`` is null.
```

The return type of a UDTF can also be a primitive or boxed type. A primitive return type indicates

> the function will never return ``null``, where as a boxed type indicates it may return ``null``.

Reviewer: ```suggestion
the function will never return ``null``, but a boxed type indicates that it may return ``null``.
```

> The KSQL server will check the value being passed to each parameter and report an error to the server

Reviewer: ```suggestion
KSQL checks the value being passed to each parameter and reports an error to the server
```

log for any null values being passed to a primitive type. The associated column in the output row
will be ``null``.

Dynamic return type
~~~~~~~~~~~~~~~~~~~

UDTFs support dynamic return types that are resolved at runtime. This is useful if you want to
implement a UDTF with a non-deterministic return type. A UDTF which returns ``BigDecimal``,
for example, may vary the precision and scale of the output based on the input schema.

To use this functionality, you need to specify a method with signature
``public SqlType <your-method-name>(final List<SqlType> params)`` and annotate it with ``@SchemaProvider``.
Also, you need to link it to the corresponding UDF by using the ``schemaProvider=<your-method-name>``
parameter of the ``@Udtf`` annotation.

> Please note that if your UDTF method returns a value of type ``List<T>`` then the type referred to

Reviewer: ```suggestion
If your UDTF method returns a value of type ``List<T>``, the type referred to
```

by the schema provider method is the type ``T`` not the type ``List<T>``.

.. _example-udtf-class:

Example UDTF class
~~~~~~~~~~~~~~~~~~

The class below creates a UDTF named ``split_string``. The name of the UDTF is provided in the ``name``

> parameter of the ``UdtfDescription`` annotation. This name is case-insensitive and is what can be

Reviewer: ```suggestion
parameter of the ``UdtfDescription`` annotation. This name is case-insensitive, and you can
```

> used to call the UDTF.

Reviewer: ```suggestion
use it to call the UDTF.
```

UDTF methods must return a value of type ``List<T>`` where T is any of the supported KSQL Java types.

> As can be seen this UDTF can be invoked in two different ways:

Reviewer: ```suggestion
You can invoke this UDTF in two different ways:
```

- with a single String containing the String to split
- with a String containing the String to split and a regex to define the delimiter

.. code:: java

    import io.confluent.ksql.function.udf.Udtf;
    import io.confluent.ksql.function.udf.UdtfDescription;

    @UdfDescription(name = "split_string", description = "splits a string into words")
    public class SplitString {

      @Udtf(description="Splits a string into words")
      public List<String> split(String input) {
        return Arrays.asList(String.split("\\s+"));
      }

      @Udtf(description="Splits a string into words")
      public List<String> split(String input, String delimRegex) {
        return Arrays.asList(String.split(delimRegex));
      }

    }

If you're using Gradle to build your UDF or UDAF, specify the ``ksql-udf``
dependency:

.. codewithvars:: bash

    compile 'io.confluent.ksql:ksql-udf:|release|'

To compile with the latest version of ``ksql-udf``:

.. codewithvars:: bash

    compile 'io.confluent.ksql:ksql-udf:+'

> If you're using Maven to build your UDF or UDAF, specify the ``ksql-udf``

Reviewer: ```suggestion
If you're using Maven to build your function, specify the ``ksql-udf``
```
